### ISSUE TITLE
Long freeze with man command on MacOS

### ISSUE THREAD

**Issue by nix4lyfe** [zdharma-continuum/fast-syntax-highlighting#27]
When I use the `man` command and hit space there is a 3-5 second pause before any input is registered, it happens irrespective of terminal emulator used. Man seems to be the only command it happens with. I am using the current version of MacOS.

**Comment by LambdaP** [comment]
This is also something I experience, using macOS 12.3.1.

**Comment by steakknife** [comment]
Happens here too, making FSH unusable with major pauses.

- Terminal: iTerm2 latest
- macOS: Catalina latest 10.15.7 (19H1922) 
- zsh: 5.9

**Comment by steakknife** [comment]
I would look to see if it's an interaction with completion.

**Comment by steakknife** [comment]
I switched to [zsh-syntax-highlighting](https://github.com/zsh-users/zsh-syntax-highlighting) because it works for me.

**Comment by LambdaP** [comment]
> I switched to [zsh-syntax-highlighting](https://github.com/zsh-users/zsh-syntax-highlighting) because it works for me.

That’s what I ended up doing as well.

**Comment by nkhlmn** [comment]
I'm also having this issue (macos 12.4)

**Comment by strawhat-dev** [comment]
As a workaround just create a quick wrapper function (can't be named `man`),
```bash
help() { man "$@"; }
```

Just creating an alias (e.g. `alias help='man'`) also did not work and still causes the terminal to freeze.

If you don't care about installing another package, you could use [batman](https://github.com/eth-p/bat-extras/blob/master/doc/batman.md) instead. Now you have colored man pages 🌈.

**Comment by LeuschkeTressa** [comment]
Same problem here (macOS): after `man`, there are intermittent lags (1-3 seconds) when continuing to type.

This happens both on Terminal.app and iTerm2 (3.4.16).

versions:
zsh 5.8.1 (x86_64-apple-darwin21.0)
macOS 12.5.1 (21G83)

**Comment by LeuschkeTressa** [comment]
If this helps someone else, here is a workaround. It will disabling syntax highlighting for commands starting with `man`.

Edit `fast-syntax-highlighting.plugin.zsh`. For its path, check the `source` command of `.zshrc`.

Look for the function `_zsh_highlight()`, around line 61, and add the following line after the first statement (`local ret=$?`) of the function:

```zsh
  [[ ${BUFFER:0:3}  == "man" ]] && return $ret
```

That is, the top of the function should now read:

```zsh
[...]

_zsh_highlight()
{
  # Store the previous command return code to restore it whatever happens.
  local ret=$?

  [[ ${BUFFER:0:3}  == "man" ]] && return $ret  # <-- added for 'man' bugfix

  # Remove all highlighting in isearch, so that only the underlining done by zsh itself remains.
  # For details see FAQ entry 'Why does syntax highlighting not work while searching history?'.
  if [[ $WIDGET == zle-isearch-update ]] && ! (( $+ISEARCHMATCH_ACTIVE )); then

[...]
```

This works for me.

**Comment by LeuschkeTressa** [comment]
I did som digging on this.

# Workaround

Anywhere in `.zshrc`, add this line:
```zsh
function whatis() { if [[ -v THEFD ]]; then :; else command whatis "$@"; fi; }
```

**Explanation:**

The lag is caused by calls to `whatis`, which the highlighter calls when highlighting `whatis` and `man` commands. `whatis` typically takes a few seconds or so to execute, so the terminal will be blocked for that time.

The line above declares a function, which will have precedence over the system `whatis` command. The function returns immediately if called from within the syntax highlighter (by testing if variable `THFD` exists, which happens to be present in the syntax highlighter where the `whatis` calls are made). Otherwise, e.g. when calling `whatis` from the terminal, the function dispatches the system `whatis` command.

# Some more info for anyone trying to fix this

The `whatis` calls are made in file `/:chroma/-whatis.ch`, function `-fast-whatis-chroma-callback()`, as `exec {THEFD}< <( [...] LANG=C whatis [...] )`. This command doesn't block though. It however stores a file descriptor for the output of the command to `$THEFD`. This file descriptor is read from in file/function `-fast-zts-read-all` with the `sysread` command - _this_ is the call that will actually block.

An easy quick-fix would be to simply disable syntax highlighting of the `man` and `whatis` commands for macOS.

**Comment by lucax88x** [comment]
unfortunately this plugin is REALLY slow in OSX, even with a m1pro max.

when I try to input something like 

`git checkout branch/`

it starts freezing for full seconds

The workaround from @LeuschkeTressa did not fix it.

**Comment by Aloxaf** [comment]
Another workround is disable highlight for man command

```
FAST_HIGHLIGHT[chroma-man]=
```

**Comment by londbell** [comment]
> unfortunately this plugin is REALLY slow in OSX, even with a m1pro max.
> 
> when I try to input something like
> 
> `git checkout branch/`
> 
> it starts freezing for full seconds
> 
> The workaround from @LeuschkeTressa did not fix it.

+1

**Comment by j-lakeman** [comment]
Same issue on up-to-date macOS 11.6.8

**Comment by j-lakeman** [comment]
> 


> Another workround is disable highlight for man command
> 
> ```
> FAST_HIGHLIGHT[chroma-man]=
> ```


Is that to be exported in `.zprofile`?
Or just set as a variable in `.zshrc`?

**Comment by mojolo** [comment]
This issue also occurs on Ubuntu varieties (I'm using Kubuntu 23.04).
👍 to @LeuschkeTressa - his workaround does fix the freezing for me.

**Comment by LeuschkeTressa** [comment]
@lucax88x , @londbell 
My workaround is only expected to work for freezing behaviour when colorizing the `man` commands (and as a consequence, the `whatis` command).

For any significant lags when other commands are colorized, there is probably something else that `fast-syntax-highlighting` does upon colorizing that specific command that is blocking for some time.

To solve such problems, you could try to disable syntax highlighting entirely for that specific command, e.g. trying the suggestion by @Aloxaf above.

## For anyone wanting to find a better fix

For anyone wanting to trouble-shoot lag when highlighting other commands than `man`/`whatis`, to try to find a workaround that doesn't involve disabling highlighting for that command, this is the best solution I found:

```zsh

# new shell without loading the typical zsh startup files:
zsh -f

# manually load the highlighter (you might need to change the path):
_zfst_file="/opt/homebrew/opt/zsh-fast-syntax-highlighting/share/zsh-fast-syntax-highlighting/fast-syntax-highlighting.plugin.zsh"
[[ -f "$_zfst_file" ]] && source "$_zfst_file" && fast-theme -q free

# load zsh's profiler module:
zmodload zsh/zprof

# here: start typing the misbehaving command, so the lag/blocking behaviour is triggered. Wait it out. 
man cow  # replace with misbehaving command

# save profiling results:
zprof > zprof.txt

```
Copy-paste the above to the terminal without the comments, otherwise `zsh` will yield an error for the `#`'s since we're using the default shell options.

Now check `zprof.txt`. It will list the executed functions ordered by execution time, with the longest-running first: this will have been numbered `1)` by `zprof`. The function indicated there will probably be some top-level function, so check the breakdown of `1)` slightly further down. There you'll see the specific highligter function that caused the lag.

E.g. for the lagging `man` command:
```bash
$ grep ' 1)' -B 2 zprof.txt

# output:
num  calls                time                       self            name
-----------------------------------------------------------------------------------
 1)    2        9045.14  4522.57   99.72%   9045.14  4522.57   99.72%  -fast-zts-read-all
--

       2/2      9045.14  4522.57   99.72%   9045.14  4522.57             -fast-whatis-chroma-callback [6]
 1)    2        9045.14  4522.57   99.72%   9045.14  4522.57   99.72%  -fast-zts-read-all
```

So: the lag (9 seconds) happened in the function `-fast-zts-read-all`, who's break-down shows that it was the call to `-fast-whatis-chroma-callback` that caused the lag. Hence, further investigation/profiling should be done on this function.

**Comment by lukejans** [comment]
Yup, this issue seems to have something to do with the `whatis` command and also the `man` command which calls the `whatis` command. I'll attach a picture of the process breakdown from `btop` which shows the root of the problem. There is a zombie process that is causing 800% cpu usage which has a child `makewhatis` process.

<img width="606" alt="Image" src="https://github.com/user-attachments/assets/646d2dda-cb48-4a0f-a976-6d262747fc61" />

**Comment by vegerot** [comment]
I'm having the same issue too on Debian

@lukejans how did you create that profile?

**Comment by carlwr** [comment]
@lukejans yes. To colorize commands taking names of man pages as arguments, the highlighter calls `whatis` to learn whether such a man page exist or not. The command line buffer will be blocked ("frozen") until that `whatis` call returns. With a correctly configured man page system, `whatis` returns immediately and no freeze will be experienced.

However, on macOS with the default configuration from Apple, `whatis` will rebuild the man page database each time it is called, and that takes some number of second. `whatis` calls out to `makewhatis` to rebuild the database, and hence that will show up as the hog-ed process.

As a side note, for me, with the latest macOS, this problem is still present. The workaround I descreibed in https://github.com/zdharma-continuum/fast-syntax-highlighting/issues/27#issuecomment-1267278072 still fixes the problem though.
```

### Somew further details on the macOS case for the interested reader

`makewhatis` is not on `PATH`; `whatis` calls it with its path `/usr/libexec/makewhatis.local` hard-coded. Both `whatis` and `makewhatis` are shell scripts so they can be inspected. All of this is from an old man page system of FreeBSD, at a time when FreeBSD used `mandoc` as the manpage viewer and these custom scripts for other parts of the man page system. Apple has made a number of patches/hacks to these scripts to get them working reasonably. FreeBSD has left this implementation since long and is now using (as most or all of the BSDs) the evolved `mandoc` not only as the manpage viewer, but for the rest of the man page system, including e.g. `whatis`, `makewhatis` etc. (A few Linux distros use `mandoc` as well, but most use another man system implementation, `mandb`.)

**Comment by lukejans** [comment]
> I'm having the same issue too on Debian
> 
> [@lukejans](https://github.com/lukejans) how did you create that profile?

I was just using `btop` but @carlwr had a much better example [here](https://github.com/zdharma-continuum/fast-syntax-highlighting/issues/27#issuecomment-1666894790) using the zsh profiler.

**Comment by carlwr** [comment]
As a hint to @mojolo, @vegerot, and anyone else having these issues, **regardless of OS/distro**:

Any lagging/freezing for commands including names of man pages, e.g. `man`, is likely caused by `whatis` not returning immediately when invoked. That can be verified easily enough by just running it, with any argument, e.g

```sh
whatis whatis
  # whatis (1)           - display one-line manual page descriptions
```
, and noticing whether it returns immediately or with some lag.

Any lag is a sign of some problem with the configuration of the man page system: `whatis` will search some pre-built database or cache if present, and if not present, will typically build it before returning which likely takes some time. With a mis-configured system, the built database/cache is likely not stored properly and as a consequence it will be rebuilt each time `whatis` is invoked.

If `whatis` is indeed the culprit, possible fixes include [disabling highlighting](https://github.com/zdharma-continuum/fast-syntax-highlighting/issues/27#issuecomment-1329248704) of the troubled commands entirely, or using [the workaround suggested by me](https://github.com/zdharma-continuum/fast-syntax-highlighting/issues/27#issuecomment-1267278072). The latter should, and has been reported to, work on Linux distros as well.
